content/client-lib-development-guide/features.textile

```diff
@@ -309,7 +336,16 @@ h3(#realtime-channel). Channel
-** If token auth is being used along with a @clientId@, the @clientId@ will be present on any messages published. A test should exist that connects with a @clientId@, publishes a message on a channel, and with another client, the message should be received with the correct @clientId@ attribute of the message
+** Unauthenticated clients without a @clientId@ (any @clientId@ is permitted):
+*** When a @Message@ with a @clientId@ value is published, Ably will accept and publish that message with the provided @clientId@. A test should assert that the @clientId@ of the published @Message@ is correct for both authentication schemes (token and basic)
+** Authenticated clients with a @clientId@ (as a result of either an explicitly configured @clientId@ in @ClientOptions@, or implicitly through token auth):
+*** When publishing a @Message@ with the @clientId@ attribute set to @null@:
+**** It is unnecessary for the client to set the @clientId@ of the @Message@ before publishing
+**** Ably is expected to automatically assign a @clientId@ upon receiving the @Message@. A test should assert that the @clientId@ value is populated for the @Message@ when received
+*** When publishing a @Message@ with the @clientId@ attribute value set to the authenticated client's @clientId@, Ably will accept the message and publish it. A test should assert that the @clientId@ value is populated for the @Message@ when received
+*** When publishing a @Message@ with a different @clientId@ attribute value to the authenticated @clientId@, the client library should reject publishing immediately. The message should not be sent to Ably and an exception should be raised. The connection and channel remains available for further operations
+*** When using token auth with a literal token, an auth URL or an auth callback, until authentication has completed, the client library will not be constrained to a @clientId@. As such, the @Auth#clientId@ attribute will be @null@ until authentication is complete. Therefore, if a @Message@ with a @clientId@ value is published before authentication has completed, the client library should not reject the @clientId@ value and the message with the specified @clientId@ should be delivered to the Ably system. A test should publish a message before authentication with a valid @clientId@ to ensure it is received back with the @clientId@ intact, and then a further test should publish a message with a @clientId@ value that does not match the @clientId@ following auth, which should be rejected by Ably when delivered and the error should contain the server error message. The connection and channel remains available for further operations
```

Comment:
As above, with realtime a message will be queued until authentication has happened, whereas with rest it will publish immediately after authentication has happened.